### PR TITLE
[dashboard] New, add statsd incr to the API

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

Choose one

- [ ] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Adds statsd counters to all GET methods and custom POST, PUT, DELETE, bulk_delete, export dashboard endpoints. Also adds statds timing metric, so all API endpoints report the time required to execute

```
DEBUG:superset.stats_logger:[stats_logger] (incr) DashboardRestApi.get_list.success
DEBUG:superset.stats_logger:[stats_logger] (timing) DashboardRestApi.get_list.time | 0.013094598001771374
127.0.0.1 - - [13/Apr/2020 17:33:10] "GET /api/v1/dashboard/?q={%22order_column%22:%22dashboard_title%22,%22order_direction%22:%22asc%22,%22page%22:0,%22page_size%22:25} HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [13/Apr/2020 17:33:10] "GET /api/v1/dashboard/?q={%22order_column%22:%22dashboard_title%22,%22order_direction%22:%22asc%22,%22page%22:0,%22page_size%22:25} HTTP/1.1" 200 -
DEBUG:superset.stats_logger:[stats_logger] (incr) DashboardRestApi.get_list.success
DEBUG:superset.stats_logger:[stats_logger] (timing) DashboardRestApi.get_list.time | 0.012102608001441695
127.0.0.1 - - [13/Apr/2020 17:33:11] "GET /api/v1/dashboard/?q={%22order_column%22:%22dashboard_title%22,%22order_direction%22:%22desc%22,%22page%22:0,%22page_size%22:25} HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [13/Apr/2020 17:33:11] "GET /api/v1/dashboard/?q={%22order_column%22:%22dashboard_title%22,%22order_direction%22:%22desc%22,%22page%22:0,%22page_size%22:25} HTTP/1.1" 200 -
DEBUG:superset.stats_logger:[stats_logger] (incr) DashboardRestApi.get_list.success
DEBUG:superset.stats_logger:[stats_logger] (timing) DashboardRestApi.get_list.time | 0.012785728999006096
127.0.0.1 - - [13/Apr/2020 17:33:12] "GET /api/v1/dashboard/?q={%22order_column%22:%22dashboard_title%22,%22order_direction%22:%22asc%22,%22page%22:0,%22page_size%22:25} HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [13/Apr/2020 17:33:12] "GET /api/v1/dashboard/?q={%22order_column%22:%22dashboard_title%22,%22order_direction%22:%22asc%22,%22page%22:0,%22page_size%22:25} HTTP/1.1" 200 -
```

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@willbarrett

### CODE REVIEW

File: superset/dashboards/api.py [hunk 1]
```diff
@@ -160,19 +160,25 @@ def post(self) -> Response:
             500:
               $ref: '#/components/responses/500'
         """
+        self.incr_stats("init", self.post.__name__)
```
Comment (craig-rueda) [inline superset/dashboards/api.py:163]: Can we dry this up? Why not just increment inside of `response_xxx` and set the method name in `@expose`?

Comment (dpgaspar) [reply superset/dashboards/api.py:163]: Good idea! (this is making things a bit uglier)

**Comment by willbarrett** [comment]
Should we add tests for this functionality?

**Comment by dpgaspar** [comment]
@willbarrett, totally, added tests

**Review by willbarrett — APPROVED** [review]
Thanks for adding tests!

**Review by villebro — COMMENTED** [review]
Looking robust! But could be even more robust with that timer decorator 🙂 Thoughts?

File: superset/views/base_api.py [hunk 2]
```diff
@@ -36,16 +36,28 @@
...
 
 
+def time_function(func: Callable, *args, **kwargs) -> Tuple[float, Any]:
+    """
+        Measures the amount of time a function takes to execute in ms
+    :param func: The function execution time to measure
+    :param args: args to be passed to the function
+    :param kwargs: kwargs to be passed to the function
+    :return: A tuple with the duration and response from the function
+    """
```
Comment (villebro) [inline superset/views/base_api.py:46]:
```suggestion
def time_function(func: Callable, *args, **kwargs) -> Tuple[float, Any]:
    """
    Measures the amount of time a function takes to execute in ms

    :param func: The function execution time to measure
    :param args: args to be passed to the function
    :param kwargs: kwargs to be passed to the function
    :return: A tuple with the duration and response from the function
    """
```

**Review by villebro — CHANGES_REQUESTED** [review]
Should `time_function` be placed in utils to make it generally available? Also a small docstring nit.

**Review by villebro — CHANGES_REQUESTED** [review]
Sorry, I missed the other docstrings. I'm being picky here, but based on PEP-257, the convention for single line headlines appears to be like this:
image
or like this:
image

Similarly multiline docstrings should be formatted like this:
image

**Comment by dpgaspar** [comment]
oh! I've been doing it wrong :(

File: tests/dashboards/api_tests.py [hunk 3]
```diff
@@ -121,14 +121,23 @@ def test_get_dashboard(self):
         db.session.delete(dashboard)
         db.session.commit()
 
+    def test_info_dashboard(self):
+        """
+            Dashboard API: Test info
+        """
```
Comment (villebro) [inline tests/dashboards/api_tests.py:127]: Still one to go

**Review by villebro — CHANGES_REQUESTED** [review]
Last nit

File: tests/dashboards/api_tests.py (hunk 3 above)
Comment (dpgaspar) [reply tests/dashboards/api_tests.py]: Not one a bunch of them :) fixed all docs string for dashboards

**Review by villebro — APPROVED** [review]
LGTM, sorry for derailing this into a nit-PR  🛤️

**Comment by dpgaspar** [comment]
It's better now, thanks for the review @villebro

**Pull request merged**